hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/checksum/ECBlockChecksumComputer.java

```diff
@@ -121,66 +119,40 @@ private void computeCompositeCrc() throws IOException {
-    // (default = 16KB) the last checksum in an EC chunk is only generated for
-    // offset.
-    long bytesPerCrcOffset = chunkSize % bytesPerCrc;
-
     long keySize = keyInfo.getDataSize();
-    // Total parity checksum bytes per stripe to remove
-    int parityBytes = getParityBytes(chunkSize, bytesPerCrc);
-
-    // Number of checksum per chunk, Eg: 2MB EC chunk will
-    // have 2 checksum per chunk.
-    int numChecksumPerChunk = (int)
-        (Math.ceil((double) chunkSize / bytesPerCrc));
 
     CrcComposer blockCrcComposer =
         CrcComposer.newCrcComposer(dataChecksumType, bytesPerCrc);
 
     for (ContainerProtos.ChunkInfo chunkInfo : chunkInfoList) {
       ByteString stripeChecksum = chunkInfo.getStripeChecksum();
+      long chunkSize = chunkInfo.getLen();
+
+      // Total parity checksum bytes per stripe to remove
+      int parityBytes = getParityBytes(chunkSize, bytesPerCrc);
 
       Preconditions.checkNotNull(stripeChecksum);
       final int checksumSize = stripeChecksum.size();
       Preconditions.checkArgument(checksumSize % 4 == 0,
           "Checksum Bytes size does not match");
-      CrcComposer chunkCrcComposer =
-          CrcComposer.newCrcComposer(dataChecksumType, bytesPerCrc);
 
       // Limit parity bytes as they do not contribute to fileChecksum
       final ByteBuffer byteWrap = stripeChecksum.asReadOnlyByteBuffer();
       byteWrap.limit(checksumSize - parityBytes);
 
-      long chunkOffsetIndex = 1;
       while (byteWrap.hasRemaining()) {
-
-        /*
-        When chunk size is not a multiple of bytes.per.crc we get an offset.
-        For eg, RS-3-2-1524k is not a multiple of 1MB. So two checksums are
-        generated 1st checksum for 1024k bytes and 2nd checksum for 500k bytes.
-        When we reach the 2nd Checksum we need to modify the bytesPerCrc as in
-        this case 500k is the bytes for which the checksum is generated.
-        */
-        long currentChunkOffset = Long.MAX_VALUE;
-        if ((chunkOffsetIndex % numChecksumPerChunk == 0)
-            && (bytesPerCrcOffset > 0)) {
-          currentChunkOffset = bytesPerCrcOffset;
+        // Here Math.min in mainly required for last stripe's last chunk. The last chunk of the last stripe can be
+        // less than the chunkSize, chunkSize is only calculated from each stripe's first chunk. This would be fine
+        // for rest of the stripe because all the chunks are of the same size. But for the last stripe we don't know
+        // the exact size of the last chunk. So we calculate it with the of keySize. If the key size is smaller
+        // than the chunk size, then we know it is the last stripe' last chunk.
+        long remainingChunkSize = Math.min(keySize, chunkSize);
+        while (byteWrap.hasRemaining() && remainingChunkSize > 0) {
+          final int checksumData = byteWrap.getInt();
+          blockCrcComposer.update(checksumData, Math.min(bytesPerCrc, remainingChunkSize));
```

Comment:
If I understand correctly, this should handle the case when chunk size is not a multiple of bytes.per.crc right ? as we have removed the code pertaining to this part above.

Reply:
I am not sure I understand this section.

Also, is this routine called on a block? If an EC key has multiple blocks, then first the checksums are combined for block 1 in a call to `computeMd5Crc` and then a subsequent call will be made for block 2 to combine its checksums into the value for block 1? How does the keySize work then across multiple blocks?

Reply:
@sodonnel Thanks for the catch. It should be `blockSize` instead of the `keySize` as we are calculating checksum per block here.